ExampleMod/Content/Projectiles/ExampleWhipProjectileAdvanced.cs

```diff
@@ -0,0 +1,183 @@
+using Terraria.GameContent;
+using Terraria.ID;
+using Terraria.ModLoader;
+
+namespace ExampleMod.Content.Projectiles
+{
+	public class ExampleWhipProjectileAdvanced : ModProjectile
+	{
+		// The texture doesn't have the same name as the item, so this property points to it.
+		public override string Texture => "ExampleMod/Content/Projectiles/ExampleWhipProjectile";
+
+		public override void SetStaticDefaults() {
+			// This makes the projectile use whip collision detection and allows flasks to be applied to it.
+			ProjectileID.Sets.IsAWhip[Type] = true;
+		}
+
+		public override void SetDefaults() {
+			Projectile.width = 18;
+			Projectile.height = 18;
+			Projectile.friendly = true;
+			Projectile.penetrate = -1;
+			Projectile.tileCollide = false;
+			Projectile.ownerHitCheck = true; // This prevents the projectile from hitting through solid tiles.
+			Projectile.extraUpdates = 1;
+			Projectile.usesLocalNPCImmunity = true;
+			Projectile.localNPCHitCooldown = -1;
+		}
+
+		private float Timer {
+			get => Projectile.ai[0];
+			set => Projectile.ai[0] = value;
+		}
+
+		private float ChargeTime {
+			get => Projectile.ai[1];
+			set => Projectile.ai[1] = value;
+		}
+
+		public override void AI() {
+			Player owner = Main.player[Projectile.owner];
+			Projectile.rotation = Projectile.velocity.ToRotation() + MathHelper.PiOver2; // Without PiOver2, the rotation would be off by 90 degrees counterclockwise.
+
+			Projectile.Center = Main.GetPlayerArmPosition(Projectile) + Projectile.velocity * Timer;
+			// Vanilla uses Vector2.Dot(Projectile.velocity, Vector2.UnitX) here. Dot Product returns the difference between two vectors, 0 meaning they are perpendicular.
+			// However, the use of UnitX basically turns it into a more complicated way of checking if the projectile's velocity is above or equal to zero on the X axis.
+			Projectile.spriteDirection = Projectile.velocity.X >= 0f ? 1 : -1;
+
+			Charge();
+
+			float swingTime = owner.itemAnimationMax * Projectile.MaxUpdates;
+
+			if (Timer >= swingTime || owner.itemAnimation <= 0) {
+				Projectile.Kill();
+				return;
+			}
+
+			owner.heldProj = Projectile.whoAmI;
+
+			// These two lines ensure that the timing of the owner's use animation is correct.
+			owner.itemAnimation = owner.itemAnimationMax - (int)(Timer / Projectile.MaxUpdates);
+			owner.itemTime = owner.itemAnimation;
+
+			if (Timer == swingTime / 2) {
+				// Plays a whipcrack sound at the tip of the whip.
+				List<Vector2> points = Projectile.WhipPointsForCollision;
+				Projectile.FillWhipControlPoints(Projectile, points);
+				SoundEngine.PlaySound(SoundID.Item153, points[points.Count - 1]);
+			}
+		}
+
+		// This method handles a charging mechanic.
+		// If you remove this, also remove Item.channel = true from the item's SetDefaults.
+		private void Charge() {
+			// Like other whips, this whip updates twice per frame (Projectile.extraUpdates = 1), so 120 is equal to 1 second.
+			if (owner.channel && ChargeTime < 120) {
```

Comment:
Where does this 'owner' variable comes from?
It seems to be missing inside of this function.

Reply:
oh so that's why it didn't build